From 93dad029bb91afbcd7efe1d34e9164cf1e89e0a2 Mon Sep 17 00:00:00 2001
From: "Srivatsa S. Bhat (VMware)" <srivatsa@csail.mit.edu>
Date: Fri, 17 Feb 2023 12:19:21 -0800
Subject: [PATCH] iavf: Fix build errors on 6.1.y

src/iavf_main.c: In function 'iavf_alloc_q_vectors':
src/iavf_main.c:1718:17: error: too many arguments to function 'netif_napi_add'

Signed-off-by: Srivatsa S. Bhat (VMware) <srivatsa@csail.mit.edu>
---
 src/iavf_main.c        |  2 +-
 src/kcompat_impl.h     | 26 ++++++++++++++++++++++++++
 src/kcompat_std_defs.h | 14 ++++++++++++++
 3 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/src/iavf_main.c b/src/iavf_main.c
index 779e752..32ec01d 100644
--- a/src/iavf_main.c
+++ b/src/iavf_main.c
@@ -1731,7 +1731,7 @@ static int iavf_alloc_q_vectors(struct iavf_adapter *adapter)
 		cpumask_copy(&q_vector->affinity_mask, cpu_possible_mask);
 #endif
 		netif_napi_add(adapter->netdev, &q_vector->napi,
-			       iavf_napi_poll, NAPI_POLL_WEIGHT);
+			       iavf_napi_poll);
 	}
 
 	return 0;
diff --git a/src/kcompat_impl.h b/src/kcompat_impl.h
index 3890479..716f799 100644
--- a/src/kcompat_impl.h
+++ b/src/kcompat_impl.h
@@ -995,4 +995,30 @@ _kc_xsk_buff_dma_sync_for_cpu(struct xdp_buff *xdp,
 #define xsk_uses_need_wakeup xsk_umem_uses_need_wakeup
 #endif /* NEED_XSK_BUFF_POOL_RENAME */
 
+/*
+ * NEED_NETIF_NAPI_ADD_NO_WEIGHT
+ *
+ * Upstream commit b48b89f9c189 ("net: drop the weight argument from
+ * netif_napi_add") removes weight argument from function call.
+ *
+ * Our drivers always used default weight, which is 64.
+ *
+ * Define NEED_NETIF_NAPI_ADD_NO_WEIGHT on kernels 3.10+ to use old
+ * implementation. Undef for 6.1+ where new function was introduced.
+ */
+#ifdef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+static inline void
+_kc_netif_napi_add(struct net_device *dev, struct napi_struct *napi,
+		   int (*poll)(struct napi_struct *, int))
+{
+	return netif_napi_add(dev, napi, poll, NAPI_POLL_WEIGHT);
+}
+
+/* RHEL7 complains about redefines. Undef first, then define compat wrapper */
+#ifdef netif_napi_add
+#undef netif_napi_add
+#endif
+#define netif_napi_add _kc_netif_napi_add
+#endif /* NEED_NETIF_NAPI_ADD_NO_WEIGHT */
+
 #endif /* _KCOMPAT_IMPL_H_ */
diff --git a/src/kcompat_std_defs.h b/src/kcompat_std_defs.h
index a751779..1811917 100644
--- a/src/kcompat_std_defs.h
+++ b/src/kcompat_std_defs.h
@@ -31,6 +31,12 @@
 #error "KERNEL_VERSION is undefined"
 #endif
 
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,10,0))
+#else /* >= 3,10,0 */
+#define NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#endif /* 3,10,0 */
+
 /*****************************************************************************/
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(3,17,0))
 #define NEED_DEVM_KASPRINTF
@@ -302,4 +308,12 @@
 #define HAVE_NDO_FDB_DEL_EXTACK
 #endif /* 5.19.0 */
 
+
+/*****************************************************************************/
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6,1,0))
+#else /* >=6.1.0 */
+#undef NEED_NETIF_NAPI_ADD_NO_WEIGHT
+#undef HAVE_NDO_GET_DEVLINK_PORT
+#endif /* 6.1.0 */
+
 #endif /* _KCOMPAT_STD_DEFS_H_ */
-- 
2.25.1

